eval/adjust.rkt

```diff
@@ -66,53 +65,60 @@
-;   intro = get-ampls(parent)
-(define (precision-tuning ivec vregs vprecs-new varc vstart-precs)
-  (for ([instr (in-vector ivec (- (vector-length ivec) 1) -1 -1)] ; reversed over ivec
-        [n (in-range (- (vector-length vregs) 1) -1 -1)]) ; reversed over indices of vregs
-
-    (define op (car instr)) ; current operation
+; Roughly speaking, the upper precision bound is calculated as:
+;   vprecs-max[i] = (+ max-prec vstart-precs[i]), where min-prec < (+ max-prec vstart-precs[i]) < max-prec
+;   max-prec = (car (get-bounds parent))
+(define (precision-tuning ivec vregs vprecs-max varc vstart-precs vuseful)
+  (define vprecs-min (make-vector (vector-length ivec) 0))
+  (for ([instr (in-vector ivec (- (vector-length ivec) 1) -1 -1)]
+        [useful? (in-vector vuseful (- (vector-length vuseful) 1) -1 -1)]
+        [n (in-range (- (vector-length vregs) 1) -1 -1)]
+        #:when useful?)
+    (define op (car instr))
     (define tail-registers (cdr instr))
-    (define srcs (map (lambda (x) (vector-ref vregs x)) tail-registers)) ; tail of the current instr
-    (define output (vector-ref vregs n)) ; output of the current instr
-
-    (define intro (vector-ref vprecs-new (- n varc))) ; intro for the current instruction
-    (define ampls (get-ampls op output srcs)) ; ampls for the tail instructions
-
-    (define final-parent-precision
-      (max (+ intro (vector-ref vstart-precs (- n varc))) (*base-tuning-precision*)))
+    (define srcs (map (lambda (x) (vector-ref vregs x)) tail-registers))
+    (define output (vector-ref vregs n))
+
+    (define max-prec (vector-ref vprecs-max (- n varc))) ; upper precision bound given from parent
+    (define min-prec (vector-ref vprecs-min (- n varc))) ; lower precision bound given from parent
+
+    ; Final precision assignment based on the upper bound
+    (define final-precision
+      (min (max (+ max-prec (vector-ref vstart-precs (- n varc))) (*rival-min-precision*))
+           (*rival-max-precision*)))
+    (vector-set! vprecs-max (- n varc) final-precision)
+
+    ; Early stopping
+    (match (*lower-bound-early-stopping*)
+      [#t
+       (when (>= min-prec (*rival-max-precision*))
+         (*sampling-iteration* (*rival-max-iterations*)))]
+      [#f
+       (when (equal? final-precision (*rival-max-precision*))
+         (*sampling-iteration* (*rival-max-iterations*)))])
+
+    ; Precision propogation for each tail instruction
+    (define ampl-bounds (get-bounds op output srcs)) ; amplification bounds for children instructions
+    (for ([x (in-list tail-registers)]
+          [bound (in-list ampl-bounds)]
+          #:when (>= x varc)) ; when tail register is not a variable
+      (match-define (cons up-bound lo-bound) bound)
 
-    (when (>= final-parent-precision (*rival-max-precision*)) ; Early stopping
-      (*sampling-iteration* (*rival-max-iterations*)))
+      ; The minimal required precision can not be negative
+      (set! lo-bound (max 0 lo-bound))
```

Comment:
Why do we need this?

Reply:
I was thinking this way: 
In general, lower bound can be negative as well.
Negative bound has less to do with the minimal precision, since it can not be negative.
And further if we pass this negative precision - a child instruction will receive this negative `min-prec` from above and it will be considered when concluding about early stopping. 
And here is what exactly can happen:
`(- (cos (exp x) 1)`
Tuning starts with `-` and let's assume that lower bound was calculated as -10000 for `(cos ...)`.
Then, `cos` got it's own lower bound of 12000, summing up -10000+12000 we get `min-prec` for `(exp x)` 2000 - sounds like it can be evaluated in a working range. 
However, here I assume that the actual lower bound of `(exp x)` should be 12000 instead of 2000 and that negative value from `-` should not be considered in the calculation.
Therefore, by using my logic with `(set! lo-bound (max 0 lo-bound))` we may stop early and decide that `(exp x)` can not be evaluated. Without `max` operation though we would not stop.